rfcs/2021-05-23-groups2.md

```diff
@@ -0,0 +1,134 @@
+
+There may be two situations to consider:
+
+1) The recipient of this verification message can have the same message as the last one or they can have more messages received - both such scenarios are ok and do not necessarily indicate a problem, as they just might be slightly ahead in receiving the messages. They would make the last matching message as ok, and the latter messages would remain unknown - it does not indicate the lack of integrity of the group, although if the next integrity check without the same messages does (this needs to be clarified).
+2) The recipient of this verification message can have fewer messages received. This situation can be resolved in several ways:
+- try to retrieve all messages from the queue that is behind. It might work, but it may be that the sender is simply trying to send messages because the network is down.
+- wait until the next integrity check and only report integrity violation if during the next integrity check they still cannot reconcile the previous integrity check. This is probably an acceptable compromise.
+
+2. Each agent sends message receipts to sending agents - message receipts would contain a signature of the message hashes. In the agent uses the same verification key for each member of the group, as considered below, these receipts can be re-broadcasted to other agents (again, grouping them in one message to avoid `(n-1)*n` messages) as a proof that the messages were delivered.
+
+Comparing with the first approach, there are pros and cons:
+- pros:
+  - the sender would only send such integrity check messages when they have send a message to all parties, thus avoiding the situation when some messages might have been not yet sent (or failed to send and they are retrying).
+- cons:
+  - integrity check message mush contain a signature per member, so it would be substantially larger.
+  - agents must use the same verification key for all members in the group, complicating the group and connections management.
+  - in the same way as in the first scenario, some delivery notifications can be arbitrarily delayed.
+
+With either approach, each agent should probably track all the receipts of all messages from all agents. 
+
+Overall the first approach seems better. It shows who received which messages and the only case where the lack of integrity would be reported if the same message is different or some messages are skipped to some recipients (i.e. they broadcast integrity violation for some of the senders).
```

Comment:
What does it mean "same message is different"?

Reply:
same message by order number in the sequence can have different content - can be clarified.